# About Generating Permutations and Combinations
## Divide Pair Conquer
### Due: Monday, 1 February 2021, 11:59pm


There are many occasions when you need to *generate* the permutations or
combinations of a set, not just count them.

There are many algorithms for generating permutations and combinations --- you
can find them if you look.

For an application, from a biographical sketch about Donald Knuth by Kenneth
Rosen, we learn that


> "Knuth grew up in Milwaukee, where his father taught bookkeeping at a Lutheran
high school and owned a small printing business. He was an excellent student,
earning academic achievement awards. He applied his intelligence in
unconventional ways, winning a contest when he was in the eighth grade by
finding as many words as possible that could be formed from the letters in

---

> **Ziegler's Giant Bar**.

___

> This won a television set for his school and a candy bar for everyone in his class.


Knuth found over 4500 words. How many can **you** find?

In [22]:
# Import a set of english words (contains 25487 words)
!pip install english-words
from english_words import english_words_set

# Compare each word in the set to the word phrase "Ziegler's Giant Bar." 
# If the word's letters are found in the word phrase, then return true/else false.
# Map through the entire process storing the count of each true case, printing the final result.

word_phrase = {'z','i','e','g','l','e','r','s','g','i','a','n','t','b','a','r'}

# Compare letter of word to the letters of the word phrase.
# If found in the set, then return true.
def check_letter(letter): 
  if (letter in word_phrase):
    return True
  else:
    return False


# Count how many words can be made from the word phrase
word_count = 0

for word in english_words_set:
  # Count letters of the word found in the word phrase.
  letter_count = 0
  for letter in word:
    if (check_letter(letter)):
      letter_count += 1
  # If letter_count is the same as the length of the word, then it can be made.
  if (letter_count == len(word)):
    word_count += 1

print(word_count)

1085


I was only able to find 1085 words from the set of words that I used. Knuth found over 4500 words. After researching the problem, I found that the judges to the competition had only about 2000 words, which mine is less than both. I tested my solution against smaller test sets that I created. My solution was able to produce the correct number of words from each test set. I wonder what I could do to improve mine? 